connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/ClientSparkResult.scala

```diff
@@ -0,0 +1,174 @@
+ */
+package org.apache.spark.sql.connect.client
+
+import java.util.Collections
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.arrow.memory.BufferAllocator
+import org.apache.arrow.vector.FieldVector
+import org.apache.arrow.vector.ipc.ArrowStreamReader
+
+import org.apache.spark.connect.proto
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.encoders.{ExpressionEncoder, RowEncoder}
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.Deserializer
+import org.apache.spark.sql.connect.client.util.{AutoCloseables, Cleanable}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.ArrowUtils
+import org.apache.spark.sql.vectorized.{ArrowColumnVector, ColumnarBatch, ColumnVector}
+
+private[sql] class ClientSparkResult(
+    responses: java.util.Iterator[proto.ExecutePlanResponse],
+    allocator: BufferAllocator)
+    extends AutoCloseable
+    with Cleanable {
+
+  private[this] var numRecords: Int = 0
+  private[this] var structType: StructType = _
+  private[this] var encoder: ExpressionEncoder[Row] = _
+  private[this] val batches = mutable.Buffer.empty[ColumnarBatch]
+
+  private def processResponses(stopOnFirstNonEmptyResponse: Boolean): Boolean = {
+    while (responses.hasNext) {
+      val response = responses.next()
+      if (response.hasArrowBatch) {
+        val ipcStreamBytes = response.getArrowBatch.getData
+        val reader = new ArrowStreamReader(ipcStreamBytes.newInput(), allocator)
+        try {
+          val root = reader.getVectorSchemaRoot
+          if (batches.isEmpty) {
+            structType = ArrowUtils.fromArrowSchema(root.getSchema)
+            // TODO: create encoders that directly operate on arrow vectors.
+            encoder = RowEncoder(structType).resolveAndBind(structType.toAttributes)
+          }
+          while (reader.loadNextBatch()) {
+            val rowCount = root.getRowCount
+            assert(root.getRowCount == response.getArrowBatch.getRowCount) // HUH!
+            if (rowCount > 0) {
+              val vectors = root.getFieldVectors.asScala
+                .map(v => new ArrowColumnVector(transferToNewVector(v)))
+                .toArray[ColumnVector]
+              batches += new ColumnarBatch(vectors, rowCount)
+              numRecords += rowCount
+              if (stopOnFirstNonEmptyResponse) {
+                return true
+              }
+            }
+          }
+        } finally {
+          reader.close()
+        }
+      }
+    }
+    false
+  }
+
+  private def transferToNewVector(in: FieldVector): FieldVector = {
+    // Work around for ARROW-17269. If we create the transfer pair without creating the vector
+    // explicitly, MapVectors will be 'converted' into ListVectors which causes all sorts of fun
+    // bugs down the line. We need to wait until Arrow 10 is released, or until the fix is back
```

Comment:
I think we are on 10.x now do we don't need it here. If we are backporting to 12.x then we may have to add this back.